### PR TITLE
Document keepalive

### PR DESCRIPTION
**Pull request by mziolekda**

### CODE REVIEW

File: docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst [hunk 1]
```diff
@@ -325,6 +325,79 @@ Common patterns from the canton log:
...
+
+Finally, you will also see the typical GO_AWAY messages that the clients and the servers exchange when they disconnect in an organized manner.
+
+.. code-block:: none
+
+    [..] DEBUG io.grpc.netty.NettyServerHandler - [id: 0x95fab9fc, L:/127.0.0.1:10038 - R:/127.0.0.1:57250] INBOUND GO_AWAY: lastStreamId=0 errorCode=0 length=0 bytes=
+    [..] DEBUG io.grpc.netty.NettyServerHandler - [id: 0x227a18df, L:/127.0.0.1:10042 - R:/127.0.0.1:51246] OUTBOUND GO_AWAY: lastStreamId=2147483647 errorCode=0
+    length=13 bytes=6170705f726571756573746564
+
+From teh above logs you can get a pretty good idea of what happened around the moment when the connection was terminated. In particular, the absence of the ping, data and
```
Comment (pawelulewicz-da) [inline docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst:396]: A typo:

```suggestion
From the above logs you can get a pretty good idea of what happened around the moment when the connection was terminated. In particular, the absence of the ping, data and
```

File: docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst [hunk 2]
```diff
@@ -325,6 +325,79 @@ Common patterns from the canton log:
...
+can be avoided by carefully configuring the components' keepalive parameters.
+
+For the GRPC protocol stack, there are two keepalive mechanisms to consider. There is the low level TCP keep-alive feature and then there are the HTTP/2 pings.
+Canton endpoints utilize the latter. It is worthwhile getting familiar with the basics of this mechanism. The article describing
+`how to use HTTP/2 PING-based keepalives in <https://grpc.io/docs/guides/keepalive/>`_ is a good place to start.
+
+The keepalive parameters have to be set on all components involved. In this eample we will concentrate on configuring the Ledger API, similar
+approach can be used for other Canton APIs. The way to do it is described in the dedicated :ref:`keepalive article <keepalive-configuration>`.
+Next, the corresponding parameters need to be altered on the Ledger API client. Please consult the user manual specific to your application.
+The parameters on both ends should be set consistently. Special attention must be given to the fact that the keepalive time setting on the client
```
Comment (pawelulewicz-da) [inline docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst:343]: > Special attention must be given to the fact that the keepalive

Maybe let's mark this text (and possibly around it) so it's more visible (for example `:warning` or bold).

**Review by pawelulewicz-da — COMMENTED** [review]
I reviewed what I could (which is basics) being not familiar with the topic.

File: docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst [hunk 3]
```diff
@@ -325,6 +325,79 @@ Common patterns from the canton log:
...
+should not be set lower than the ``permit-keep-alive-time`` of the Ledger API server. Finally, the settings of the proxies between the client
+and the server need to be selected. Again, the user manual specific to each component is the best place to gain the knowledge on how best
+to do it. The proxy inactivity timeouts should be more lenient than than the settings of the client and the server.
+
+If you suspect that the disconnections are caused by misalignment of the keepalive settings on different components, you can troubleshoot
+the connections by enabling detailed networking logging on Ledger API server. Ideally, you would also simultaneously increase the loglevel
+in the client application. Whether you will be able to do this, may depend on the specific implementation. If the client is written on
+the java stack, the chances are that you will be able to apply similar approach to that in the Canton participant.
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst:351]:
```suggestion
You can determine if disconnections are caused by a misalingment of keep-alive settings on different components by enabling detailed networking logging on the Ledger API server. If the specific implementation allows, simultaneously increase the log level
in the client application. If the client is written on
the Java stack, you can probably apply a similar approach to that you would use in the Canton participant.
```

File: docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst [hunk 4]
```diff
@@ -325,6 +325,79 @@ Common patterns from the canton log:
...
+
+THe easiest way to get the keepalive logging going is to start Canton Participant with the ``--debug`` flag. Alternatively, you can
+modify the logger's configuration in the ``logback.xml`` file. You will have to bump the log-level of the ``io.grpc.netty.NettyServerHandler``
+the ``io.grpc.netty.NettyClientHandler`` and the ``com.digitalasset.canton.platform.apiserver.GrpcConnectionLogger``.
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst:355]:
```suggestion
The easiest way to start the keep-alive logging is to start Canton Participant with the ``--debug`` flag. Alternatively, you can
modify the logger's configuration in the ``logback.xml`` file. You then have to bump the log level of the ``io.grpc.netty.NettyServerHandler``
the ``io.grpc.netty.NettyClientHandler`` and the ``com.digitalasset.canton.platform.apiserver.GrpcConnectionLogger``.
```

File: docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst [hunk 5]
```diff
@@ -325,6 +325,79 @@ Common patterns from the canton log:
...
+
+Once that is done, you will be able to observe the life cycle events of the underlying GRPC channels such as their opening and closing:
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst:357]:
```suggestion
Once that is done, you can observe the lifecycle events of the underlying GRPC channels such as their opening and closing:
```

File: docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst [hunk 6]
```diff
@@ -325,6 +325,79 @@ Common patterns from the canton log:
...
+
+.. code-block:: none
+
+    [..] DEBUG c.d.c.p.a.GrpcConnectionLogger:participant=participant - Grpc connection open: {io.grpc.Grpc.TRANSPORT_ATTR_LOCAL_ADDR=/127.0.0.1:5001,
+    io.grpc.internal.GrpcAttributes.securityLevel=NONE, io.grpc.Grpc.TRANSPORT_ATTR_REMOTE_ADDR=/127.0.0.1:49944}
+    [..] DEBUG c.d.c.p.a.GrpcConnectionLogger:participant=participant - Grpc connection closed: {io.grpc.Grpc.TRANSPORT_ATTR_LOCAL_ADDR=/127.0.0.1:5001,
+    io.grpc.internal.GrpcAttributes.securityLevel=NONE, io.grpc.Grpc.TRANSPORT_ATTR_REMOTE_ADDR=/127.0.0.1:49944}
+
+Likewise, you will be able to observe the header and data frames being exchanged. That is the essence of the GRPC interchange.
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst:366]:
```suggestion
Likewise, you can observe the header and data frames being exchanged. That is the essence of the GRPC interchange.
```

**Review by carrielaben-da — APPROVED** [review]
Grammar changes and rewording.

**Comment by mziolekda** [comment]
Github removed all but 2 changes from the review batch. I will have to re-add them manually

File: docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst (hunk 2 above)
Comment (mziolekda) [reply docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst]: The text has been reworded and as a consequence the phrase starting with "Special attention" has been made less prominent

File: docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst [hunk 7]
```diff
@@ -325,6 +325,79 @@ Common patterns from the canton log:
 
           The log line contains the name of the missing claim, but not the actual claims. When using tokens based on user names (`Audience Based Tokens <https://docs.daml.com/app-dev/authorization.html#audience-based-tokens>`_ and `Scope Based Tokens <https://docs.daml.com/app-dev/authorization.html#scope-based-tokens>`_), consult the user management service to see whether you need to grant more rights to the user. When using tokens based on party names (`Custom Claims Access Tokens <https://docs.daml.com/app-dev/authorization.html#custom-daml-claims-access-tokens>`_), debug the token in `JWT.IO <https://jwt.io/#debugger-io>`_.
 
+Disconnections
+--------------
+
+Intermittent network failures cause the majority of disconnections that occur in real-life scenarios. Less frequently,
+disconnections happen when network proxies, load balancers, or other elements of the network infrastructure terminate the client-server
+connection because they think that there is no activity going on in the communication channel. Such disconnections are predictable and
+can be avoided by carefully configuring the components' keep-alive parameters.
```
Comment (amarqueslee-da) [inline docs/2.10.0/docs/canton/usermanual/troubleshooting_guide.rst:334]: Would recommend to be a bit careful about how the scope of the problem is represented here, as load balancers can contribute to disconnections for many and varied reasons.

For example, I worked on an issue for HKEX where a ledger client could not consume the output of `GetActiveContracts` fast enough, so backpressure from the client onto the load balancer triggered disconnection once nginx's buffer (of configurable size) was over-filled.

**Review by amarqueslee-da — COMMENTED** [review]
Very grateful to see this documentation get enriched, thank you!

**Pull request merged**